### PR TITLE
actinlib do not return from :wait-for-result, if that goal is overridden from different instance in same roseus process

### LINKED ISSUE #462: :wait-for-result does not catch result correctly when we have to action client with same name?
when there is two actionlib client for same controller it did not return from :wait-for-result, for example

```
*ri* . controller-actions
:head
:rarm
:larm
:head
```

did not work well

### PR DESCRIPTION
**Pull request by k-okada**
- add test-simple-client-cancel.test for https://github.com/start-jsk/jsk_apc/issues/2106
- print warn if :wait-for-result ends with preempted
- actinlib.l : add :name-space method to simple-action-server

Close #462

### CODE REVIEW

**Comment by k-okada** [comment]
this PR solves problem reported in https://github.com/jsk-ros-pkg/jsk_roseus/issues/462 and https://github.com/start-jsk/jsk_apc/issues/2106,
the root cause is roseus do not handle subscriber with same name https://github.com/jsk-ros-pkg/jsk_roseus/issues/407,
This PR do not deal with that situation, but only care about actionlib.

situation is 
1. we have multiple `action-client` 
2. since `roseus` only handle 1 subscriber, when result message arrives in only calls latest `action-client`
3. latest `action-client` discards arrived message for old `action-client` 

and solutions is
1. extend goal status callback queue from 1 to 8
2. when (latest) client receives message, and find it is not yours, it searches global action client list and look for who sent goal for that result message
3. change status of previous action client

File: roseus/euslisp/actionlib.l [hunk 1]
```diff
@@ -93,20 +102,24 @@
...
+                     action-result-class #'send self :action-result-cb 8)
      (ros::subscribe (format nil "~A/feedback" name-space)
-                     action-feedback-class #'send self :action-feedback-cb 1)))
+                     action-feedback-class #'send self :action-feedback-cb 8)))
    (setq simple-state ros::*simple-goal-state-done*)
+   (push self ros::*simple-action-client-list*)
    self)
+  (:dispose
+   ()
+   (setq ros::*simple-action-client-list* (remove self ros::*simple-action-client-list*)))
```
Comment (k-okada) [inline roseus/euslisp/actionlib.l]: since we push `simple-action-client` to global list, it will never disposed.... need FIX

Comment (k-okada) [inline roseus/euslisp/actionlib.l]: use something like
```
(do-all-symbols (s) (when (and (boundp s)(derivedp (eval s) ros::simple-action-client))(print s)))
```
instead?

**Comment by k-okada** [comment]
I think current implementation do not dispose created objeect, so it will have memory problem in future, I tried to write code like follows, but do not work well.

```
  (:action-result-cb
   (msg)
   (let (dummy-msg)
     (ros::ros-debug "[~A] reuslt-cb ~A ~A" name-space (send msg :status :goal_id :id) (goal-status-to-string (send msg :status :status)))
     (unless (send comm-state :update-result msg)
       (ros::ros-warn "[~A] action-result-cb may recieved old client's goal" name-space)
       (ros::ros-warn "     expected goal id ~A" (send (send comm-state :action-goal) :goal_id :id))
       (ros::ros-warn "     recieved goal id ~A" (send msg :status :goal_id :id))
       (do-all-symbols (c)
         (when (and (boundp c) (derivedp (eval c) ros::simple-action-client))
           (when (string= (send msg :status :goal_id :id)
                        (send (send (send c :comm-state) :action-goal) :goal_id :id))
             (send c :action-result-cb msg)
             (ros::ros-warn "[~A] reuslt-cb ~A -> state ~A" name-space (send msg :status :goal_id :id) (send (send c :comm-state) :state))
             )))
       (return-from :action-result-cb nil))
```

**Comment by YoheiKakiuchi** [comment]
Anonymous objects, it is not bound to symbols and can not appear in do-all-symbols.
For example, created objects on alist can not appear in do-all-symbols
```(push (instance objectclass :init) alist)```



Created instance should be bound to symbol, so
code below added to last of ```:init``` may solve this problem. (not tested ...)
```
(unless (find-package "ROSACTIONLIB") (make-package "ROSACTIONLIB"))
(set
 (gentemp "ACTIONLIBCLIENT" (find-package "ROSACTIONLIB"))
 self)
```

Package "ROSACTIONLIB" is created in order to reduce time of loop.
~~~
- (do-all-symbols (s)
+ (do-symbols (s (find-package "ROSACTIONLIB"))
~~~

**Comment by knorth55** [comment]
I tested with `DEBUG` logger level and it fails with error below.
```
Call Stack (max depth: 20):
  0: at (send (send (send ros::c :comm-state) :action-goal) :goal_id :id)
  1: at (string= (send ros::msg :status :goal_id :id) (send (send (send ros::c :comm-state) :action-goal) :goal_id :id))
  2: at (if (string= (send ros::msg :status :goal_id :id) (send (send (send ros::c :comm-state) :action-goal) :goal_id :id)) (send ros::c :action-result-cb ros::msg))
  3: at (while #:dolist26040 (setq ros::c (pop #:dolist26040)) (if (string= (send ros::msg :status :goal_id :id) (send (send (send ros::c :comm-state) :action-goal) :goal_id :id)) (send ros::c :action-result-cb ros::msg)))
  4: at (let ((ros::c nil) (#:dolist26040 ros::*simple-action-client-list*)) nil (while #:dolist26040 (setq ros::c (pop #:dolist26040)) (if (string= (send ros::msg :status :goal_id :id) (send (send (send ros::c :comm-state) :action-goal) :goal_id :id)) (send ros::c :action-result-cb ros::msg))) nil)
  5: at (dolist (ros::c ros::*simple-action-client-list*) (if (string= (send ros::msg :status :goal_id :id) (send (send (send ros::c :comm-state) :action-goal) :goal_id :id)) (send ros::c :action-result-cb ros::msg)))
  6: at (progn (ros::ros-warn "[~A] action-result-cb may recieved old client's goal" ros::name-space) (ros::ros-warn "     expected goal id ~A" (send (send ros::comm-state :action-goal) :goal_id :id)) (ros::ros-warn "     recieved goal id ~A" (send ros::msg :status :goal_id :id)) (dolist (ros::c ros::*simple-action-client-list*) (if (string= (send ros::msg :status :goal_id :id) (send (send (send ros::c :comm-state) :action-goal) :goal_id :id)) (send ros::c :action-result-cb ros::msg))))
  7: at (if (not (send ros::comm-state :update-result ros::msg)) (progn (ros::ros-warn "[~A] action-result-cb may recieved old client's goal" ros::name-space) (ros::ros-warn "     expected goal id ~A" (send (send ros::comm-state :action-goal) :goal_id :id)) (ros::ros-warn "     recieved goal id ~A" (send ros::msg :status :goal_id :id)) (dolist (ros::c ros::*simple-action-client-list*) (if (string= (send ros::msg :status :goal_id :id) (send (send (send ros::c :comm-state) :action-goal) :goal_id :id)) (send ros::c :action-result-cb ros::msg)))))
  8: at (when (not (send ros::comm-state :update-result ros::msg)) (ros::ros-warn "[~A] action-result-cb may recieved old client's goal" ros::name-space) (ros::ros-warn "     expected goal id ~A" (send (send ros::comm-state :action-goal) :goal_id :id)) (ros::ros-warn "     recieved goal id ~A" (send ros::msg :status :goal_id :id)) (dolist (ros::c ros::*simple-action-client-list*) (if (string= (send ros::msg :status :goal_id :id) (send (send (send ros::c :comm-state) :action-goal) :goal_id :id)) (send ros::c :action-result-cb ros::msg))))
  9: at (unless (send ros::comm-state :update-result ros::msg) (ros::ros-warn "[~A] action-result-cb may recieved old client's goal" ros::name-space) (ros::ros-warn "     expected goal id ~A" (send (send ros::comm-state :action-goal) :goal_id :id)) (ros::ros-warn "     recieved goal id ~A" (send ros::msg :status :goal_id :id)) (dolist (ros::c ros::*simple-action-client-list*) (if (string= (send ros::msg :status :goal_id :id) (send (send (send ros::c :comm-state) :action-goal) :goal_id :id)) (send ros::c :action-result-cb ros::msg))))
  10: at (let (ros::dummy-msg) (ros::ros-debug "[~A] reuslt-cb ~A ~A" ros::name-space (send ros::msg :status :goal_id :id) (ros::goal-status-to-string (send ros::msg :status :status))) (unless (send ros::comm-state :update-result ros::msg) (ros::ros-warn "[~A] action-result-cb may recieved old client's goal" ros::name-space) (ros::ros-warn "     expected goal id ~A" (send (send ros::comm-state :action-goal) :goal_id :id)) (ros::ros-warn "     recieved goal id ~A" (send ros::msg :status :goal_id :id)) (dolist (ros::c ros::*simple-action-client-list*) (if (string= (send ros::msg :status :goal_id :id) (send (send (send ros::c :comm-state) :action-goal) :goal_id :id)) (send ros::c :action-result-cb ros::msg)))) (ros::ros-debug "[~A] reuslt-cb ~A -> state ~A" ros::name-space (send ros::msg :status :goal_id :id) (send ros::comm-state :state)) (setq ros::dummy-msg (instance actionlib_msgs::goalstatusarray :init)) (send ros::dummy-msg :status_list (list (send ros::msg :status))) (send ros::dummy-msg :header (send ros::msg :header)) (send self :goal-status-cb ros::dummy-msg))
  11: at (ros::spin-once ros::groupname)
  12: at (ros::spin-once ros::groupname)
  13: at (cond (ros::groupname (ros::spin-once ros::groupname)) (t (ros::spin-once)))
  14: at (send self :spin-once)
  15: at (while (ros::ok) (ros::ros-debug "[~A] :wait-for-result ~A ~A" ros::name-space (ros::simple-goal-states-to-string ros::simple-state) (send ros::comm-state :state)) (if (= ros::simple-state ros::*simple-goal-state-done*) (return)) (if (> ros::timeout 0) (let* ((ros::tm (ros::time- (ros::time-now) ros::start))) (if (> (send ros::tm :to-sec) ros::timeout) (return-from :wait-for-result nil)))) (when ros::return-if-server-down (when (= (ros::get-num-publishers ros::result-topic) 0) (ros::ros-error "[~A] Unexpected returns from :wait-for-result : no publishers found for ~A" ros::name-space ros::result-topic) (return-from :wait-for-result nil)) (when (and ros::status-stamp ros::maximum-status-interval (> (send (ros::time- (ros::time-now) ros::status-stamp) :to-sec) ros::maximum-status-interval)) (ros::ros-error "[~A] Unexpected returns from :wait-for-result : status did not received for 5 seconds" ros::name-space) (return-from :wait-for-result nil))) (send self :spin-once) (ros::sleep))
  16: at (let ((ros::start (ros::time-now)) (ros::result-topic (format nil "~A/result" ros::name-space))) (ros::ros-debug "[~A] :wait-for-result ~A, timeout=~A" ros::name-space (if ros::goal_id (send ros::goal_id :id)) ros::timeout) (unless ros::goal_id (ros::ros-error "[~A] :wait-for-result (return nil when no goal exists)" ros::name-space) (return-from :wait-for-result nil)) (ros::rate 10) (while (ros::ok) (ros::ros-debug "[~A] :wait-for-result ~A ~A" ros::name-space (ros::simple-goal-states-to-string ros::simple-state) (send ros::comm-state :state)) (if (= ros::simple-state ros::*simple-goal-state-done*) (return)) (if (> ros::timeout 0) (let* ((ros::tm (ros::time- (ros::time-now) ros::start))) (if (> (send ros::tm :to-sec) ros::timeout) (return-from :wait-for-result nil)))) (when ros::return-if-server-down (when (= (ros::get-num-publishers ros::result-topic) 0) (ros::ros-error "[~A] Unexpected returns from :wait-for-result : no publishers found for ~A" ros::name-space ros::result-topic) (return-from :wait-for-result nil)) (when (and ros::status-stamp ros::maximum-status-interval (> (send (ros::time- (ros::time-now) ros::status-stamp) :to-sec) ros::maximum-status-interval)) (ros::ros-error "[~A] Unexpected returns from :wait-for-result : status did not received for 5 seconds" ros::name-space) (return-from :wait-for-result nil))) (send self :spin-once) (ros::sleep)) (ros::ros-debug "[~A] :wait-for-result finished ~A" ros::name-space (ros::goal-status-to-string (send self :get-state))) (if (eq (send self :get-state) actionlib_msgs::goalstatus::*preempted*) (ros::ros-warn "[~A] :wait-for-result finished with preempted status" ros::name-space)) (eq (send self :get-state) actionlib_msgs::goalstatus::*succeeded*))
  17: at (send-all controller-actions :wait-for-result :timeout timeout)
  18: at (send-all controller-actions :wait-for-result :timeout timeout)
  19: at (send-all controller-actions :wait-for-result :timeout timeout)
  And more...
/home/shingo/jskeus/eus/Linux64/bin/irteusgl 0 error: cannot find method :goal_id in (send (send (send ros::c :comm-state) :action-goal) :goal_id :id)
```
DEBUG log: https://gist.github.com/knorth55/c8e76f8d5ce62877665259e790d25661
See https://github.com/start-jsk/jsk_apc/issues/2106, too.

File: roseus/euslisp/actionlib.l [hunk 2]
```diff
@@ -56,7 +57,15 @@
...
-     (send comm-state :update-result msg)
+     (unless (send comm-state :update-result msg)
+       (ros::ros-warn "[~A] action-result-cb may recieved old client's goal" name-space)
+       (ros::ros-warn "     expected goal id ~A" (send (send comm-state :action-goal) :goal_id :id))
+       (ros::ros-warn "     recieved goal id ~A" (send msg :status :goal_id :id))
+       ;; look for exiting objects make from simple-goal-client
+       (dolist (c ros::*simple-action-client-list*)
+         (if (string= (send msg :status :goal_id :id)
+                      (send (send (send c :comm-state) :action-goal) :goal_id :id))
+             (send c :action-result-cb msg))))
```
Comment (knorth55) [inline roseus/euslisp/actionlib.l]: https://github.com/jsk-ros-pkg/jsk_roseus/pull/519#issuecomment-303914148

**Comment by knorth55** [comment]
We checked this PR with the new commit solve the problem.
We use baxter in both simulator and real world. https://github.com/start-jsk/jsk_apc/issues/2106

**Pull request merged**